### PR TITLE
Support ENV vars and console parameters in application-level - Closes #2976

### LINKED ISSUE #2976: Support ENV vars and command line arguments in application-level
### Expected behavior
With recent changes the configuration is distributed to each module and components. So as the environment variables should move along to individual configuration.

### Actual behavior
 Currently we have a central place for all the definition of environment variables (old helpers/config.js)

### Which version(s) does this affect? (Environment, OS, etc...)
1.6.0

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?
Env variables definition was on central file, not compatible with new approach of modular configurations. 

### How did I fix it?
By making env variable and command line arguments support to be part of the schema itself. 

### How to test it?
From README file check the list of variables and try to set those individually. 

### Review checklist

* The PR resolves #2976
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

**Comment by nazarhussain** [comment]
@MaciejBaj This PR is ready for review and following points to be kept in mind: 

1. `PROTOCOL_VERSION` - Added only for tests no config is schema attached to it
2. `config.coverage` - Removed as I can't find nay reference usage 
3. `--inspect-workers, --inspect-brokers` - Removed as I can't find any reference usage and its not bounded to any configuration as well
4. `config.api.mode = 'READONLY'` Setting this value during snapshotting is removed, because I also can't find any reference usage for `api.mode`
5. Reset of the options to disable syncing and broadcasting is moved chain module rather the configuration 
```
appConfig.network.enabled = false;
appConfig.network.list = [];
appConfig.broadcasts.active = false;
appConfig.syncing.active = false;
```


#### Things to do 

1. Now using `LISK_PEERS`, ` --peers` is required to specify the port along with ips. In old implementation the ports were required in config but was optional in the env variable and command line option. If you wan to keep the old behaviour then better is to adjust the schema so the port stay optional for `modules.chain.network.list` and then fill in ports with `modules.chain.network.wsPort` if not exists. 
2. Need to write some replacement logic for old `helpers/config.js` for `update_config.js`, it was only used for validation in that file. The script will work even without validation and application will validate it during the startup. Need to think on it further as configuration compilation logic is now distributed between modules so would not suggest to use it in the network level implementation script.

File: framework/src/controller/helpers/validator/keywords/index.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+const env = require('./env');
+const arg = require('./arg');
+
+module.exports = { env, arg };
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/index.js]: ❤️

File: package.json [hunk 2]
```diff
@@ -89,6 +89,7 @@
 		"tempy": "0.2.1",
 		"valid-url": "1.0.9",
 		"wamp-socket-cluster": "2.0.0-beta.4",
+		"yargs": "=13.2.1",
```
Comment (shuse2) [inline package.json]: no need for `=`

File: framework/src/controller/helpers/validator/keywords/arg.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
+const debug = require('debug')('lisk:validator:arg');
+const commandLineArguments = require('yargs').argv;
+const _ = require('lodash');
+const formatters = require('./formatters');
+
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/arg.js]: prefer to put `validator` at the root level

File: src/index.js [hunk 4]
```diff
@@ -1,32 +1,69 @@
...
+	},
+};
 
 try {
+	const { NETWORK, CUSTOM_CONFIG_FILE } = validator.validateWithDefaults(
+		appSchema,
+		{}
+	);
+
+	/* eslint-disable import/no-dynamic-require */
```
Comment (shuse2) [inline src/index.js]: prefer not to use `dynamic-require`.
Please use `fs` instead

File: framework/src/controller/helpers/validator/keywords/formatters.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+				return {
+					ip,
+					wsPort: wsPort || 5000,
+				};
+			});
+		}
+		return [];
+	},
+
+	stringToDelegateList: value => {
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/formatters.js]: i think this is not framework logic

File: framework/src/controller/helpers/validator/keywords/env.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	if (typeof schema === 'string') {
+		envVariableName = schema;
+		variableFormatter = null;
+	} else {
+		envVariableName = schema[0];
+		variableFormatter = formatters[schema[1]] || null;
+	}
+
+	return function(data, dataPath, object, key) {
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/env.js]: prefer to use consistently anonymous function

File: framework/src/controller/helpers/validator/keywords/env.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+			maxItems: 2,
+		},
+	],
+};
+
+const compile = (schema, parentSchema) => {
+	debug('compile: schema: %j', schema);
+	debug('compile: parent schema: %j', parentSchema);
+
+	let envVariableName;
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/env.js]: prefer to use `const`

File: framework/src/controller/helpers/validator/keywords/arg.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+		return [option, alias];
+	}
+
+	return [];
+};
+
+const compile = (schema, parentSchema) => {
+	debug('compile: schema: %j', schema);
+	debug('compile: parent schema: %j', parentSchema);
+	let argNames;
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/arg.js]: prefer to use const

File: framework/src/controller/helpers/validator/keywords/arg.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+	if (typeof schema === 'string') {
+		argNames = extractArgsNames(schema);
+		argsFormatter = null;
+	} else if (Array.isArray(schema)) {
+		argNames = extractArgsNames(schema[0]);
+		argsFormatter = formatters[schema[1]] || null;
+	}
+
+	return function(data, dataPath, object, key) {
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/arg.js]: prefer to consistently use anonymous function

File: src/index.js [hunk 10]
```diff
@@ -39,5 +76,5 @@ try {
 			process.exit();
 		});
 } catch (e) {
-	console.error(e);
+	console.error(e.errors[1]);
```
Comment (shuse2) [inline src/index.js]: where does it actually throw and why it's index 1?

File: framework/src/controller/helpers/validator/keywords/arg.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		let argValue;
+
+		argNames.forEach(argName => {
+			if (!argValue) {
+				argValue = commandLineArguments[argName] || undefined;
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/arg.js]: why we cannot use native function here?

File: framework/src/controller/helpers/validator/index.js [hunk 12]
```diff
@@ -15,6 +16,8 @@ const validatorWithDefaults = new Ajv({
 	useDefaults: true,
 	$data: true,
 });
+validatorWithDefaults.addKeyword('env', envKeyword);
+validatorWithDefaults.addKeyword('arg', argKeyword);
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/index.js]: why do we need these why do we need these keywords?

File: framework/src/modules/http_api/defaults/config.js [hunk 13]
```diff
@@ -26,10 +30,12 @@ const DefaultConfig = {
 				public: {
 					type: 'boolean',
 					default: false,
+					env: 'LISK_API_PUBLIC',
 				},
 				whiteList: {
 					type: 'array',
 					default: ['127.0.0.1'],
+					env: ['LISK_API_WHITELIST', 'stringToIpPortSet'],
```
Comment (shuse2) [inline framework/src/modules/http_api/defaults/config.js]: i think env and formatter should be separate property.
ie)
env: 'LISK_API_WHITELIST',
parser: 'stringToIPPortSet',
Also default doesn't have port setup

File: framework/src/controller/helpers/validator/keywords/formatters.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
+module.exports = {
```
Comment (shuse2) [inline framework/src/controller/helpers/validator/keywords/formatters.js]: formatting is not validation imo

File: framework/src/controller/helpers/validator/index.js (hunk 12 above)
Comment (MaciejBaj) [reply framework/src/controller/helpers/validator/index.js]: I will add an explanation in JSDoc:
```
/**
 * Adds a possibility to validate the app parameters for Lisk Framework provided through:
 * 1. Environment variables (env) like "npm start LISK_NETWORK=test"
 * 2. Command line arguments (arg) like "npm start --network test"
 * These keywords extend Ajv Validator and follow the format:
 * @type {{compile, errors, modifying, valid, metaSchema}}
 */
```

File: framework/src/controller/helpers/validator/keywords/formatters.js (hunk 5 above)
Comment (nazarhussain) [reply framework/src/controller/helpers/validator/keywords/formatters.js]: Chain module uses `LISK_FORGING_DELEGATES` env variable which can only be specified as string. So this logic is responsible to convert string to delegate list.

File: framework/src/modules/http_api/defaults/config.js (hunk 13 above)
Comment (nazarhussain) [reply framework/src/modules/http_api/defaults/config.js]: Passing parser as separate key to parent schema applies it to json config as well. If you want we can make `env` key it be an object and then specify parser.

File: framework/src/controller/helpers/validator/keywords/formatters.js (hunk 14 above)
Comment (nazarhussain) [reply framework/src/controller/helpers/validator/keywords/formatters.js]: Yes but its a step covert values to right format before validation.

**Comment by nazarhussain** [comment]
@shuse2 @MaciejBaj @fchavant Please keep in mind this PR will introduce a breaking change to configuration. 

https://github.com/LiskHQ/lisk/pull/2977#issuecomment-468978151

Now using `LISK_PEERS`, `--peers` is required to specify the port along with ips. In old implementation the ports were required in config but was optional in the env variable and command line option. If you wan to keep the old behaviour then better is to adjust the schema so the port stay optional for modules.chain.network.list and then fill in ports with `modules.chain.network.wsPort` if not exists.

File: framework/src/controller/application.js [hunk 15]
```diff
@@ -86,36 +87,49 @@ class Application {
 		constants = {},
 		config = { components: { logger: null }, modules: {} }
 	) {
-		if (typeof label === 'function') {
-			label = label.call();
+		let appConfig;
+
+		// If user passes multiple config objects merge them in left-right order
+		if (Array.isArray(config)) {
```
Comment (shuse2) [inline framework/src/controller/application.js]: Constructor config is initializing as object and treating as array is bit strange.
Please use consistent type

File: framework/src/controller/helpers/validator/keywords/formatters.js (hunk 14 above)
Comment (shuse2) [reply framework/src/controller/helpers/validator/keywords/formatters.js]: I think default and formatting are still not concern of validator.
I will create a discussion issue to gather opinion and if necessary we will change it later.
Also, I will create an issue to use @liskhq/lisk-validator

**Comment by nazarhussain** [comment]
@shuse2 Apparently these seems different things, but we are using JSON-Schema specification which is very rich in feature set. It not just contains `default` but also `description` and `examples` keywords as well. 

https://json-schema.org/understanding-json-schema/reference/generic.html

So any tool like `Ajv` can't be called only validators rather than libraries implementing the specification. 

The topic you raise here is are we fully using a specification standard or subset of it. To me its much easier to communicate and establish the concept that we are using **JSON-Schema 7.0** specification and we use all of its features provided.

**Comment by shuse2** [comment]
@nazarhussain Good point on the latest JSON schema spec.
For this PR, i think it's fine to include default/formatter, but let's discuss this with @MaciejBaj 

I think it's fine to have it in the file, but using the validator to modify the value 🤔

**Comment by nazarhussain** [comment]
So we have separate validator method `validateWithDefaults` for this purpose, standard method `validate` does not modify the values and don't use defaults to validate.

File: framework/src/controller/application.js (hunk 15 above)
Comment (nazarhussain) [reply framework/src/controller/application.js]: It was intended to pass multiple custom config files. Remember for application network config or custom config are both custom configuration. 

This particular part will be removed with the PR of https://github.com/LiskHQ/lisk/issues/3171 where we have config compilation logic separated and application will accept unified compile configuration.

File: scripts/update_config.js [hunk 16]
```diff
@@ -393,7 +394,10 @@ history.migrate(
 		// Set the env variables to validate against correct network and config file
 		process.env.LISK_NETWORK = network;
 		process.env.LISK_CONFIG_FILE = tempFilePath;
-		new AppConfig(packageJSON, false);
+
+		// TODO: Add a replacement lgoic for config helper
```
Comment (limiaspasdaniel) [inline scripts/update_config.js]: `logic`

File: package.json [hunk 17]
```diff
@@ -105,6 +105,7 @@
 		"tempy": "0.2.1",
 		"valid-url": "1.0.9",
 		"wamp-socket-cluster": "2.0.0-beta.4",
+		"yargs": "13.2.2",
```
Comment (MaciejBaj) [inline package.json]: Only yargs should be in this PR - ` feature/improve_overall_configuration` is not up to date.

File: framework/src/controller/helpers/validator/keywords/arg/index.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+					formatter: formatters[schema.formatter] || null,
+				};
+
+	return function(data, dataPath, object, key) {
+		let argValue;
+		const commandLineArguments = yargs.argv;
+
+		argVariable.names.forEach(argName => {
+			if (!argValue) {
+				argValue = commandLineArguments[_.camelCase(argName)] || undefined;
```
Comment (MaciejBaj) [inline framework/src/controller/helpers/validator/keywords/arg/index.js]:
```suggestion
				// Remove "-" or "--" from command line argument names
				argValue = commandLineArguments[_.camelCase(argName)] || undefined;
```

File: package.json (hunk 17 above)
Comment (nazarhussain) [reply package.json]: The only addition is this PR is `yargs`, rest were formatted by `npm install`

Comment (shuse2) [reply package.json]: I see only `yargs` included in this PR, so im resolving this

**Pull request merged**